### PR TITLE
Fix zero size allocation in new_delete_resource

### PR DESCRIPTION
**Pull request by jrhemstad**
fixes https://github.com/thrust/thrust/issues/1052

I attempted to add a new unit test for this, but I'm not familiar with building/running Thrust's unit tests, so I can't verify if I did it correctly.

### CODE REVIEW

File: testing/mr_new.cu [hunk 1]
```diff
@@ -33,4 +33,16 @@ void TestNewDeleteResourceAlignedAllocation()
         }
     }
 }
+
+void TestNewDeleteResourceEmptyAllocation()
+{
+    thrust::mr::new_delete_resource memres;
+    ASSERT_EQUAL(nullptr, memres.do_allocate(0));
+    memres.do_deallocate(nullptr,0);
```
Comment (jrhemstad) [inline testing/mr_new.cu]: What's the canonical way in Thrust to test that this succeeds?

File: thrust/mr/new.h [hunk 2]
```diff
@@ -65,6 +69,11 @@ class new_delete_resource THRUST_FINAL : public memory_resource<>
         ::operator delete(p, bytes, std::align_val_t(alignment));
 #else
         (void)alignment;
+        
+        // If `p` is null, do nothing
+        if(nullptr == p){
```
Comment (jrhemstad) [inline thrust/mr/new.h]: I wonder if this should also check for `bytes == 0`?

Comment (allisonvacanti) [reply thrust/mr/new.h]: I'm assuming that the value of `bytes` must be the same in `do_deallocate` as it was when `do_allocate` was called, in which case there's no need to do the additional check. Adding `assert((nullptr == p) == (0 == bytes));` might be a nice diagnostic for debug builds, though.

@griwes Do you agree?

File: testing/mr_new.cu (hunk 1 above)
Comment (griwes) [reply testing/mr_new.cu]: This seems good.

File: testing/mr_new.cu [hunk 3]
```diff
@@ -33,4 +33,16 @@ void TestNewDeleteResourceAlignedAllocation()
...
+}
+
+
+
 DECLARE_UNITTEST(TestNewDeleteResourceAlignedAllocation);
+DECLARE_UNITTEST(TestNewDeleteResourceEmptyAllocation);
```
Comment (griwes) [inline testing/mr_new.cu]: Make sure that the `DECLARE_UNITTEST` is always the next line after the closing brace of the function it uses to declare the test.

File: thrust/mr/new.h [hunk 4]
```diff
@@ -43,6 +43,10 @@ class new_delete_resource THRUST_FINAL : public memory_resource<>
 #if __cplusplus >= 201703L
         return ::operator new(bytes, std::align_val_t(alignment));
 #else
+        // don't allocate anything if the caller requested zero bytes
+        if(0 == bytes){
+           return nullptr;
```
Comment (griwes) [inline thrust/mr/new.h]: I will need to consult someone who understands the C++ library spec better as to whether we are allowed to do this or not. I'm not entirely sure if we are.

File: testing/mr_new.cu (hunk 1 above)
Comment (allisonvacanti) [reply testing/mr_new.cu]: From what I can tell, this should be fine. The only way `do_deallocate` can signal failure is to throw an exception, which will be caught by the unit test framework.

File: thrust/mr/new.h (hunk 2 above)
Comment (griwes) [reply thrust/mr/new.h]: It's a precondition that you invoke `deallocate` with the same arguments that were passed to the `allocate` call that produced the pointer. It's fine for us to just do nothing on nullptr, assuming returning it is correct in the first place (see the comment above); it's UB if the user does something wrong here, so just always doing nothing on nullptr would be perfectly fine.

Comment (griwes) [inline thrust/mr/new.h]: `if (!p)<line break>{`, please.

**Comment by allisonvacanti** [comment]
This looks good to me, I just want to make sure my assumption about the `bytes` parameter is correct.

Can you squash this branch into a single commit to simply our CI testing?

File: thrust/mr/new.h [hunk 5]
```diff
@@ -43,6 +43,10 @@ class new_delete_resource THRUST_FINAL : public memory_resource<>
 #if __cplusplus >= 201703L
         return ::operator new(bytes, std::align_val_t(alignment));
 #else
+        // don't allocate anything if the caller requested zero bytes
+        if(0 == bytes){
```
Comment (griwes) [inline thrust/mr/new.h]: `if (bytes == 0)<line break>{`, please.

**Comment by jrhemstad** [comment]
Talking with @griwes it turns out that the existing implementation for `do_allocate` was already doing the right thing. 

Relevant line(s) in the standard: http://eel.is/c++draft/basic.stc.dynamic.allocation#2.sentence-5

> If the request succeeds, the value returned by a replaceable allocation function is a non-null pointer value ([conv.ptr]) p0 different from any previously returned value p1, unless that value p1 was subsequently passed to a replaceable deallocation function.

...

> The effect of indirecting through a pointer returned from a request for zero size is undefined.

... 

> C++ differs from C in requiring a zero request to return a non-null pointer. 


So it would appear it is actually required that `new_delete_resource::allocate` does _not_ return `nullptr` more than once. 

That leaves the question of whether or not `deallocate` should check for `nullptr`. Since it is legal for `allocate` to return `nullptr` at least once, then `deallocate` should also check for this.

**Comment by jrhemstad** [comment]
As @griwes points out, it's illegal to pass anything into `deallocate` that wasn't returned by `allocate`. As such, if `allocate` never returns `nullptr`, then `deallocate` doesn't need to check for it!

Closing.

**Pull request closed without merging**